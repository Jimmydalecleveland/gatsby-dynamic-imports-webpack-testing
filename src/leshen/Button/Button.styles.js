import styled from '@emotion/styled'
import { css } from '@emotion/core'
import { stripUnit, rem } from 'polished'

export const buttonStyles = ({
  color,
  theme,
  rounded,
  shadow,
  variant,
  outlined,
  ghost,
  leftIcon,
  rightIcon,
}) => {
  let borderRadius
  let colorAdjustment = 'dark'

  if (rounded === 'min') {
    borderRadius = theme.radius
  } else if (rounded === 'max') {
    borderRadius = '2em'
  } else {
    borderRadius = 'none'
  }

  if (!theme.colors[color].dark) {
    colorAdjustment = 'light'
  }

  return css`
    display: flex;
    align-items: center;
    justify-content: center;
    padding: ${theme.spacing.xxs} ${theme.spacing.sm};
    font: inherit;
    color: ${theme.colors[color].base.readable};
    background-color: ${theme.colors[color].base.hex};
    border: 0;
    border-radius: ${borderRadius};
    cursor: pointer;
    transition: all 0.3s;
    -webkit-appearance: button;

    > * {
      pointer-events: none;
    }

    .button-text {
      ${leftIcon && `margin-left: ${theme.spacing.xxs}`};
      ${rightIcon && `margin-right: ${theme.spacing.xxs}`};
    }

    &:hover {
      color: ${theme.colors[color][colorAdjustment].readable};
      background-color: ${theme.colors[color][colorAdjustment].hex};
    }

    &:active {
      color: ${theme.colors[color][`${colorAdjustment}er`].readable};
      background-color: ${theme.colors[color][`${colorAdjustment}er`].hex};
    }

    /* TODO: determine design system specifications for this state */
    &:disabled {
      background-color: ${theme.colors.light.darkest.hex};
      color: ${theme.colors.dark.lightest.hex};
    }

    ${shadow && `box-shadow: ${theme.shadow}`};

    ${
      variant === 'small' &&
      `
      padding: ${theme.spacing.min} ${theme.spacing.xxs};
      `
    };

    ${
      ghost &&
      `
      color: ${theme.colors[color].base.hex};
      box-shadow: none;
      background-color: transparent;
      `
    }

    ${
      variant === 'feature' &&
      `
      padding-top: ${theme.spacing.xs};
      padding-bottom: ${theme.spacing.xs};

      .button-text {
        margin-left: ${leftIcon ? theme.spacing.xs : ''};
        margin-right: ${rightIcon ? theme.spacing.xs : ''};
      }
      `
    }

    ${
      outlined &&
      `
      background-color: transparent;
      color: ${theme.colors[color].base.hex};
      border: 2px solid ${theme.colors[color].base.hex};

      &:hover {
        background-color: transparent;
        color: ${theme.colors[color][colorAdjustment].hex};
        border-color: ${theme.colors[color][colorAdjustment].hex};
      }

      &:active {
        background-color: transparent;
        color: ${theme.colors[color][`${colorAdjustment}er`].hex};
        border-color: ${theme.colors[color][`${colorAdjustment}er`].hex};
      }
    `
    };

    ${
      variant === 'mega' &&
      css`
        padding: ${theme.spacing.sm} ${theme.spacing.max};
        justify-self: stretch;
        flex-direction: column;

        .button-text {
          margin-top: ${leftIcon && theme.spacing.sm};
          margin-bottom: ${rightIcon && theme.spacing.sm};
          margin-left: 0;
          margin-right: 0;
        }
      `
    }

    ${
      variant === 'hero' &&
      `
      padding-top: ${theme.spacing.sm};
      padding-bottom: ${theme.spacing.sm};
      padding-left: ${
        leftIcon
          ? rem(stripUnit(theme.spacing.xl) - stripUnit(theme.spacing.min))
          : theme.spacing.xl
      };
      padding-right: ${
        rightIcon
          ? rem(stripUnit(theme.spacing.xl) - stripUnit(theme.spacing.min))
          : theme.spacing.xl
      };

      .button-text {
        margin-left: ${leftIcon ? theme.spacing.sm : ''};
        margin-right: ${rightIcon ? theme.spacing.sm : ''};
      }
    `
    }
  `
}

export const Button = styled.button(buttonStyles)
