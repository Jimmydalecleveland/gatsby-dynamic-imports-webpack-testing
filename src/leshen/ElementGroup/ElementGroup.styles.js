import styled from '@emotion/styled'

/* TODO: resolve emotions "":first-child is potentially unsafe" issue.
 ** Check for updates to this Github issue: https://github.com/emotion-js/emotion/issues/1105#issuecomment-491304409
 ** And this issue for more explanation of the problem: https://github.com/emotion-js/emotion/issues/1178
 */
export const ElementGroup = styled('div')`
  display: inline-flex;

  > button {
    margin: 0;
  }

  > button:first-child {
    border-top-right-radius: 0;
    border-bottom-right-radius: 0;
  }

  > button:last-child {
    border-top-left-radius: 0;
    border-bottom-left-radius: 0;
  }

  > button:not(:first-child):not(:last-child) {
    border-radius: 0;
  }
`
