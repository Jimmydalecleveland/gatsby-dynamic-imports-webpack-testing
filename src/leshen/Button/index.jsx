import React from 'react'
import PropTypes from 'prop-types'

import Typography from '../Typography'
import * as Styled from './Button.styles'

const Button = ({
  children,
  variant,
  ghost,
  leftIcon,
  rightIcon,
  color,
  ...props
}) => {
  let textVariant = null

  if (variant === 'small') {
    textVariant = 'small'
  } else if (variant === 'feature') {
    textVariant = 'feature'
  } else if (variant === 'hero') {
    textVariant = 'h5'
  } else if (variant === 'mega') {
    textVariant = 'feature'
  }

  return (
    <Styled.Button
      variant={variant}
      leftIcon={leftIcon}
      rightIcon={rightIcon}
      color={color}
      ghost={ghost}
      tabIndex="0"
      {...props}
    >
      {leftIcon}
      <Typography
        className="button-text"
        variant={textVariant}
        rendersAs="span"
      >
        {children}
      </Typography>
      {rightIcon}
    </Styled.Button>
  )
}

Button.defaultProps = {
  rounded: 'min',
  shadow: true,
  outlined: false,
  ghost: false,
  type: 'button',
  color: 'primary',
  className: null,
  variant: null,
  leftIcon: null,
  rightIcon: null,
}

Button.propTypes = {
  children: PropTypes.node.isRequired,
  variant: PropTypes.oneOf(['default', 'small', 'feature', 'hero', 'mega']),
  rounded: PropTypes.oneOf(['none', 'min', 'max']),
  shadow: PropTypes.bool,
  outlined: PropTypes.bool,
  /** Uses color for text, and makes background invisible. Hover states still apply original styles. */
  ghost: PropTypes.bool,
  type: PropTypes.oneOf(['button', 'submit', 'reset']),
  color: PropTypes.oneOf(['primary', 'secondary', 'tertiary', 'dark', 'light']),
  className: PropTypes.string,
  leftIcon: PropTypes.node,
  rightIcon: PropTypes.node,
}

export default Button

export const ticketID = 'CMS-158'
