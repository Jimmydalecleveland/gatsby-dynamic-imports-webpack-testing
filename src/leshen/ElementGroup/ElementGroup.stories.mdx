import { Story, Meta, Preview, Props, Anchor } from '@storybook/addon-docs/blocks'

import ElementGroup, { ticketID } from '.'
import Button from '../Button'
import ComponentProgress from '../ComponentProgress'

<Meta title='Components|ElementGroup' component={ElementGroup} decorators={[(story) => <><ComponentProgress ticketID={ticketID} />{story()}</>]} />

# ElementGroup
Takes input and button children and styles them so they fit snuggly together.
<Preview>
  <Story name="basic">
    <ElementGroup>
      <input type="text" placeholder="Input" />
      <Button>Search</Button>
    </ElementGroup>
  </Story>
</Preview>

## Props
<Props of={ElementGroup} />

## Double Capped
<Preview>
  <Story name="double capped">
    <ElementGroup>
      <Button>Add</Button>
      <input type="text" placeholder=". . ." />
      <Button>Subtract</Button>
    </ElementGroup>
  </Story>
</Preview>

## Multiple Buttons
<Preview>
  <Story name="multiple buttons">
    <ElementGroup>
      <input type="text" placeholder="Input" />
      <Button color="dark">Cancel</Button>
      <Button color="primary">Accept</Button>
    </ElementGroup>
  </Story>
</Preview>
