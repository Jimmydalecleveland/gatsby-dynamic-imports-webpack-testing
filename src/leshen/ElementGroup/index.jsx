import React from 'react'
import PropTypes from 'prop-types'

import * as Styled from './ElementGroup.styles'

const ElementGroup = ({ children, ...others }) => (
  <Styled.ElementGroup {...others}>{children}</Styled.ElementGroup>
)

ElementGroup.propTypes = {
  /** Allows anything to be passed, but only styles inputs and buttons are supported */
  children: PropTypes.node.isRequired,
}

export default ElementGroup

export const ticketID = 'CMS-161'
