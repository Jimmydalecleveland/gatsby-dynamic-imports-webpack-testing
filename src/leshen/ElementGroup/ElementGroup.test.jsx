import React from 'react'
import { render } from 'setup-tests'
import { CacheProvider } from '@emotion/core'
import createCache from '@emotion/cache'

import ElementGroup from '.'
import Button from '../Button'

/* This is some voodoo to get rid of the console.error from emotion:
 * `The pseudo class ":first-child" is potentially unsafe when doing server-side rendering`
 * We actually need to use :first-child (or haven't figured out a way around it) so we'll ignore this for now
 */
const emotionCache = createCache()
emotionCache.compat = true

describe('Link', () => {
  test('Link should render', () => {
    const { container } = render(
      <CacheProvider value={emotionCache}>
        <ElementGroup>
          <input type="text" placeholder="Input" />
          <Button>Search</Button>
        </ElementGroup>
      </CacheProvider>
    )
    expect(container.firstChild).toMatchSnapshot()
  })
})
