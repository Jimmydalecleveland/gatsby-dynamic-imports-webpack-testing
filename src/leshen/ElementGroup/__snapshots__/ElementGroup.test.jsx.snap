// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link Link should render 1`] = `
.emotion-4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-4 > button {
  margin: 0;
}

.emotion-4 > button:first-child {
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
}

.emotion-4 > button:last-child {
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
}

.emotion-4 > button:not(:first-child):not(:last-child) {
  border-radius: 0;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 8px 16px;
  font: inherit;
  color: #fff;
  background-color: #0f6bc0;
  border: 0;
  border-radius: 4px;
  cursor: pointer;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  -webkit-appearance: button;
  box-shadow: 0 0.3rem 1rem 0 rgba(0,0,0,0.1);
}

.emotion-2 > * {
  pointer-events: none;
}

.emotion-2 .e1u83bah1 svg,
.emotion-2 svg {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  fill: #fff;
  height: 16px;
  width: 16px;
}

.emotion-2:hover {
  color: #fff;
  background-color: #125b9f;
}

.emotion-2:active {
  color: #fff;
  background-color: #134b7f;
}

.emotion-2:disabled {
  background-color: #cacfd4;
  color: #5b636d;
}

.emotion-0 {
  font-size: 1rem;
  line-height: 1.375rem;
}

@media all and (-ms-high-contrast:none),(-ms-high-contrast:active) {
  .emotion-0 {
    width: 100%;
  }
}

<div
  class="emotion-4 emotion-5"
>
  <input
    placeholder="Input"
    type="text"
  />
  <button
    class="emotion-2 emotion-3"
    color="primary"
    tabindex="0"
    type="button"
  >
    <span
      class="button-text emotion-0 emotion-1"
    >
      Search
    </span>
  </button>
</div>
`;
